docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -137,38 +170,97 @@ that account fields can be retrieved for signing.
 
 #### Interface `oneof` Handling
 
-If the module needs to work with any `sdk.Msg`s that use interface types, that
-`sdk.Msg` should be implemented as an interface with getters and setters on the
-module level and a no-arg constructor function should be passed around to
-required CLI and REST client commands.
+If a module needs to work with `sdk.Msg`s that use interface types, those interface
+types should be dealt with using `google.protobuf.Any` for signing and a `oneof`
+at the app-level for encoding.
+
+Using `google.protobuf.Any` for signing will allow client libraries to create
+reusable code for dealing with interface types that doesn't require regenerating
+protobuf client libraries for every chain. Using `oneof`s at the encoding level
+saves space in the Tendermint block store.
+
+Modules should define a generic `sdk.Msg` for interface types at the module
+level that uses `Any`. Ex:
+
+```go
+// x/gov/types/types.proto
+message MsgSubmitAnyProposal {
```

Comment:
This will most likely cause confusion. I'd recommend coming up with better nomenclature. This is only used for client-size tx signing.

Reply:
How about `MsgSubmitGenericProposal` or simply `MsgSubmitProposal`? Just noting that this is likely the type clients will interact with most (as opposed to the app-level version)

Reply:
Right. Maybe we use `Client` and/or `Sign` in the name?

Reply:
Well from a client's perspective it's not actually any different from any other `Msg` (so adding `Client or `Sign` might be confusing). For most developers, I imagine it will be the primary way to interact with modules (as opposed to the app `oneof`). So why not just `MsgSubmitProposal`?

Reply:
Because that's even more confusing IMHO. The name should perhaps be indicative of what it's being used for -- which is signing.

Reply:
I'd opt to use `ClientMsgSubmitProposal` and have a VERY clear godoc stating its purpose and use.

Reply:
But all of the types are client types from the client's perspective so while single this type out. They may not even know or care about the app-level encoding msg.

And it's not just signing, but also tx submission and likely tx decoding (haven't mentioned that yet but I imagine block explorers will want to use this variant). The app level version may likely only get used for merkle proofs by clients.

If anything I would argue we prefix the app-level types with either `App` or `Encoding` or something. These msg's should just look like any other Msg for clients, maybe with `Any` or `Generic` in the name somewhere.

Anyway if we can't settle on something, I'd suggest we either get some other opinions or address in our architecture meeting Thurs.

Reply:
This is already confusing the mess out of me. How do you expect others to keep track of this?

> And it's not just signing, but also tx submission and likely tx decoding (likely block explorers will want to use this variant).

Huh? I presumed this was solely for tx sig generation. The app-level types are used to encode the tx and broadcast it.

Reply:
I will write some details to the ADR to help make this more clear. What I am referring to is a convenience gRPC endpoint that abstracts over tx submission and decoding - like our current `/tx` REST endpoints. I guess this should be in this ADR. Hopefully that will help

Reply:
See this commit diff: https://github.com/cosmos/cosmos-sdk/pull/6031/files/11fc3f5be33623971ec35015d0819ceec54daca3..2d0efb4e5247f5f1de5ec39de04c3d1cc8fad154

I don't actually think it's that complicated but I can see why it's a bit confusing.

Basically I'm saying let's provide clients with a way to generically interact with apps and modules, and insulate them from the messy details of app-specific `oneof`s. All that complexity will be handled on the backend. So I think for it will be a lot simpler for clients this way and much less of a breaking change.

Of course, there is a level of complexity on the backend needed to deal with all the app-specific `oneof`s but that's not really new.

Reply:
I think I know what you're getting, so see my new comments on the topic.